Improve filename handling in cbmcopy

thierer opened the pull request:
This PR fixes the issues discussed in #79:

- `cbmread` replaces even slashes in the filename given by the "-o" option with underscores. This makes it impossible to write the output to anywhere but the current directory (c8c5f07).
- `cbmread` adds an extension that reflects the CBM filetype ("`.prg`", "`.seq`", ...), but `cbmwrite` doesn't remove it and neither sets the filetype accordingly (b40f163).

It also addresses these additional issues I found while looking into this:

- A slash isn't the only problematic character that is valid for a filename in CBM DOS. [According to Wikipedia](https://en.wikipedia.org/wiki/Comparison_of_file_systems#Limits), `\`, `"`, `<`, `>` and `|` are illegal for FAT and NTFS as well. And while they are legal for ext2/3/4, they are still annoying even on Linux, as the shell treats them as special characters. That's why I don't differentiate between Windows and Linux (044b8b5).
- While we're at it, also replace characters that are illegal for CBM DOS with a space in `cbmwrite` (ac47a80).
- `cbmwrite` uses the filename in PETSCII format for its status message ("writing x -> y"). This displays garbage when the filename contains uppercase PETSCII characters, as they are not valid ASCII characters. So make a temporary copy of the filename, convert it to ASCII and use that for the status message (35e143f).

Some remarks:

- `cbmread` also creates files with a "`.del`" extension (is it even possible to read them?). As I wasn't able to create them, even by using a "D" filetype, I didn't bother to handle them when writing.
- Some of these changes are not backwards compatible and might have an negative impact (especially GUI front ends come to mind):
  - c8c5f07 and 044b8b5 because files might now end up with a different name and even in a different directory than expected.
  - b40f163 because a file might end up on the floppy drive with the extension removed and a different filetype. 
  - 35e143f because it changes the case of both the filename and the filetype and operation characters in the status message. This could be changed, but I'd argue that the new version is correct.
- Tested on Linux, but not on Windows.

spiro-trikaliotis commented:
Hello Martin,
my time is limited at the moment. I am trying to understand what exactly the original code does and how your fix changes the behaviour.
To be more precise, I want to understand *why* the original code does what it does. At the moment, I do not understand the reasoning. So, I fear I might make a mistake if I commit your changes, because I might introduce a problem that I do not understand yet.

Of course, it might be that the code was wrong in the first place, and there is no reasoning behind it...

spiro-trikaliotis commented:
I must admit I cannot completely follow the logic behind your changes.
Can you give some good test examples where the behaviour before and afterwards changes?

thierer commented:
### Setup

```
> ls -FR
.:
out_dir/

./out_dir:

> cbmctrl dir 8
0 ."cbmcopy         " md 2a
17   "testfile"         prg   
647 blocks free.             
00, ok,00,00
```

### 1. Tests with current master ff18f4de

a.  As discussed in #79, when a local output filename given by the `-o` option references a relative or absolute path, the file is still written to the current directory, with all path separators replaced by underscores. This is both annoying and makes it impossible to write a file anywhere but the current directory. Fixed by c8c5f07.
```
> cbmcopy -r -o out_dir/testfile.prg 8 testfile
[Info] reading TESTFILE -> out_dir_testfile.prg
[Info] identified a 1541-II drive
.....\[Info] 00, OK,00,00

> ls -FR
.:
out_dir/  out_dir_testfile.prg

./out_dir:
```

b. When writing files from the local filesystem to the CBM drive, characters that are illegal in CBM DOS are not replaced. They either throw an error (like '?') or are mangled (like '='). Fixed by ac47a80.
```
> cbmcopy -w -o test\?file 8 out_dir_testfile.prg
[Info] writing out_dir_testfile.prg -> TEST?FILE,P,W
[Info] identified a 1541-II drive
[Fatal] could not open file for writing: 33,SYNTAX ERROR,00,00

> cbmcopy -w -o test\=file 8 out_dir_testfile.prg
[Info] writing out_dir_testfile.prg -> TEST=FILE,P,W
[Info] identified a 1541-II drive
.....|
[Info] 00, OK,00,00

> cbmctrl dir 8
0 ."cbmcopy         " md 2a
17   "testfile"         prg   
17   "test"             prg   
630 blocks free.             
00, ok,00,00
```

c. Likewise, characters in filenames (other than a path separator) that are illegal on Windows and (while allowed) still lead to usability issues on Linux (because they have to be escaped/quoted) are not replaced when reading from the drive. Fixed by 044b8b5.
```
> cbmcopy -w -o test\<\>file 8 out_dir_testfile.prg
[Info] writing out_dir_testfile.prg -> TEST<>FILE,P,W
[Info] identified a 1541-II drive
.....|
[Info] 00, OK,00,00

> cbmctrl dir 8
0 ."cbmcopy         " md 2a
17   "testfile"         prg   
17   "test"             prg   
17   "test<>file"       prg   
613 blocks free.             
00, ok,00,00

> cbmcopy -r 8 test\<\>file
[Info] reading TEST<>FILE -> test<>file.prg
[Info] identified a 1541-II drive
.....\[Info] 00, OK,00,00

> ls -FR
.:
 out_dir/   out_dir_testfile.prg  'test<>file.prg' 

./out_dir:
```

d. As also mentioned in #79, when reading a file from the drive, it gets a file extension on the local filesystem that's based on the file type, like `.PRG`. But when writing such a file, the correct file type is neither derived from the extension (it always defaults to PRG), nor is the extension removed. Both issues are annoying when using such files from a CBM computer. Fixed by b40f163.
```
> cbmcopy -r 8 testfile
[Info] reading TESTFILE -> testfile.prg
[Info] identified a 1541-II drive
.....\[Info] 00, OK,00,00

> ls -FR
.:
 out_dir/   out_dir_testfile.prg  'test<>file.prg'   testfile.prg

./out_dir:

> cbmcopy -w 8 testfile.prg
[Info] writing testfile.prg -> TESTFILE.PRG,P,W
[Info] identified a 1541-II drive
.....|
[Info] 00, OK,00,00

> cp testfile.prg testfile.seq
> cbmcopy -w 8 testfile.seq
[Info] writing testfile.seq -> TESTFILE.SEQ,P,W
[Info] identified a 1541-II drive
.....|
[Info] 00, OK,00,00

> cbmctrl dir 8
0 ."cbmcopy         " md 2a
17   "testfile"         prg   
17   "test"             prg   
17   "testfile.prg"     prg   
17   "testfile.seq"     prg   
596 blocks free.             
00, ok,00,00
```

e. The target filename that `cbmcopy` prints when writing to the drive is already converted to PETSCII (that's why it's displayed in upper case even if the original filename is in lower case). The problem is that the PETSCII upper case characters are not equivalent to their ASCII counterparts, that's why the display is garbage when the source filename already contains upper case characters (The filename on the target drive is correct, though). Fixed by 35e143f.
```
> cbmcopy -w -o "TESTFILE" 8 testfile.prg
[Info] writing testfile.prg -> ��������,P,W
[Info] identified a 1541-II drive
.....|
[Info] 00, OK,00,00

> cbmctrl dir 8
0 ."cbmcopy         " md 2a
17   "testfile"         prg   
17   "test"             prg   
17   "testfile.prg"     prg   
17   "testfile.seq"     prg   
17   "TESTFILE"         prg   
579 blocks free.             
00, ok,00,00
```

### 2. Same tests with the patches from this PR applied:

a. Path separators are not replaced if the local filename was given as a command line option. They are still replaced in the original CBM filename.
```
> cbmcopy -r -o out_dir/testfile.prg 8 testfile
[Info] reading TESTFILE -> out_dir/testfile.prg
[Info] identified a 1541-II drive
.....\[Info] 00, OK,00,00

> ls -FR
.:
out_dir/

./out_dir:
testfile.prg

> cbmcopy -w -o test\/file 8 out_dir/testfile.prg
[Info] writing out_dir/testfile.prg -> test/file,p,w
[Info] identified a 1541-II drive
.....|
[Info] 00, OK,00,00

> cbmctrl dir 8
0 ."cbmcopy         " md 2a
17   "testfile"         prg   
17   "test/file"        prg   
630 blocks free.             
00, ok,00,00

> cbmcopy -r 8 test\/file
[Info] reading TEST/FILE -> test_file.prg
[Info] identified a 1541-II drive
.....\[Info] 00, OK,00,00

> ls -FR
.:
out_dir/  test_file.prg

./out_dir:
testfile.prg
```

b. Characters that are illegal for CBM DOS are replaced with space characters.
```
> cbmcopy -w -o test\?\=file 8 out_dir/testfile.prg
[Info] writing out_dir/testfile.prg -> test  file,p,w
[Info] identified a 1541-II drive
.....|
[Info] 00, OK,00,00

> cbmctrl dir 8
0 ."cbmcopy         " md 2a
17   "testfile"         prg   
17   "test/file"        prg   
17   "test  file"       prg   
613 blocks free.             
00, ok,00,00
```

c. Illegal/cumbersome characters are replaced with an underscore, when writing to the local filesystem.
```
> cbmcopy -w -o test\<\>file 8 out_dir/testfile.prg
[Info] writing out_dir/testfile.prg -> test<>file,p,w
[Info] identified a 1541-II drive
.....|
[Info] 00, OK,00,00

> cbmcopy -r 8 test\<\>file
[Info] reading TEST<>FILE -> test__file.prg
[Info] identified a 1541-II drive
.....\[Info] 00, OK,00,00

> ls -FR
.:
out_dir/  test__file.prg  test_file.prg

./out_dir:
testfile.prg
```

d. Any local file extension that matches a CBM DOS file type (`.PRG`, `.SEQ`, `.USR`) is removed from the target filename and the file type is set accordingly. This is the reverse operation to what is already done when reading a file.
```
> cp out_dir/testfile.prg test-prg.prg
> cp test-prg.prg test-seq.seq
> cp test-prg.prg test-usr.usr

> cbmcopy -w 8 test-prg.prg 
[Info] writing test-prg.prg -> test-prg,p,w
[Info] identified a 1541-II drive
.....|
[Info] 00, OK,00,00
> cbmcopy -w 8 test-seq.seq 
[Info] writing test-seq.seq -> test-seq,s,w
[Info] identified a 1541-II drive
.....|
[Info] 00, OK,00,00
> cbmcopy -w 8 test-usr.usr 
[Info] writing test-usr.usr -> test-usr,u,w
[Info] identified a 1541-II drive
.....|
[Info] 00, OK,00,00

> cbmctrl dir 8
0 ."cbmcopy         " md 2a
17   "testfile"         prg   
17   "test/file"        prg   
17   "test  file"       prg   
17   "test<>file"       prg   
17   "test-prg"         prg   
17   "test-seq"         seq   
17   "test-usr"         usr   
545 blocks free.             
00, ok,00,00
```

e. Target filenames with upper case characters are correctly displayed in the status message.
```
> cbmcopy -w -o "TESTFILE" 8 out_dir/testfile.prg
[Info] writing out_dir/testfile.prg -> TESTFILE,p,w
[Info] identified a 1541-II drive
.....|
[Info] 00, OK,00,00

> cbmctrl dir 8
0 ."cbmcopy         " md 2a
17   "testfile"         prg   
17   "test/file"        prg   
17   "test  file"       prg   
17   "test<>file"       prg   
17   "test-prg"         prg   
17   "test-seq"         seq   
17   "test-usr"         usr   
17   "TESTFILE"         prg   
528 blocks free.             
00, ok,00,00
```

thierer commented:
I found that the problem that a filename in PETSCII is used for the status message also exists in the `cbmread` case.

When reading back the file created in step e. above, this happens:

```
> cbmcopy -r 8 TESTFILE
[Info] reading �������� -> TESTFILE.prg
[Info] identified a 1541-II drive
.....\[Info] 00, OK,00,00
```
I just pushed 29149b7a as an updated version of 35e143f, which includes the (trivial) fix:
https://github.com/OpenCBM/OpenCBM/blob/29149b7a06e872b2299557dda2dc1dfdfbf6fb69/opencbm/cbmcopy/main.c#L615

Result with this fix:
```
> cbmcopy -r 8 TESTFILE
[Info] reading TESTFILE -> TESTFILE.prg
[Info] identified a 1541-II drive
.....\[Info] 00, OK,00,00
```

spiro-trikaliotis commented:
Merged by aa2c7fa131.
